File: apps/ruby/lib/getting_started_ml/chapter15/domain.rb
# frozen_string_literal: true

module GettingStartedMl
  # 第 15 章のドメイン層。予測の入力と、モデル・置き場の約束。HTTP にも Marshal にも依存しない。
  #
  # Ruby には trait もインターフェースも無いので、約束はメソッドの名前と戻り値の取り決めだけで表す。
  #
  # - 興行収入のモデル: predict_sales(movie) で数値を返す
  # - 生存予測のモデル: survives?(passenger) で真偽値を返す
  # - モデルの置き場: load_sales_model・load_survival_model でモデルを返す。無ければ ModelNotFound を投げる
  #
  # 約束を守っているかは、置き場の実装とテスト用の偽物に同じテスト（test/support/model_store_contract.rb）を
  # 走らせて確かめる。
  module Chapter15
    # モデルを読み込めない。モデルをまだ学習していないだけなので、API は 503 を返す。
    class ModelNotFound < StandardError
      attr_reader :model_name

      def initialize(model_name)
        @model_name = model_name
        super("モデルがありません: #{model_name}")
      end
    end

    # 映画の特徴量。SNS の評判 2 種類・主演の人気・原作の有無（0 か 1）。
    Movie = Data.define(:sns1, :sns2, :actor, :original) do
      # 第 7 章のモデルに渡す特徴量にする。
      def features
        Chapter02::Features.new(columns: Chapter07::Cinema::FEATURES, values: [sns1, sns2, actor, original].map(&:to_f))
      end
    end

    # 乗客の特徴量。第 8 章と同じく値は文字列で持ち、空文字列は欠損値として前処理に任せる。
    Passenger = Data.define(:pclass, :sex, :age, :sib_sp, :parch, :fare, :embarked) do
      # 第 8 章のパイプラインに渡す行にする。
      def row
        Chapter08::Survived.passenger([pclass, sex, age, sib_sp, parch, fare, embarked])
      end
    end

    # 第 7 章の線形回帰のモデルを、興行収入のモデルの約束に合わせる。
    LinearSalesModel = Data.define(:model) do
      def predict_sales(movie)
        model.predict_one(movie.features)
      end
    end

    # 第 8 章の学習済みパイプラインを、生存予測のモデルの約束に合わせる。
    PipelineSurvivalModel = Data.define(:pipeline) do
      def survives?(passenger)
        pipeline.predict(Chapter08::Survived.features([passenger.row])).first == Chapter08::Survived::SURVIVED
      end
    end
  end
end

File: apps/elixir/lib/getting_started_ml/chapter15/domain.ex
defmodule GettingStartedMl.Chapter15.ModelNotFoundError do
  @moduledoc """
  学習済みモデルが無いことを表す例外。

  この章のほかの失敗は `ArgumentError` のままだが、これだけは別の例外にする。
  API が 503 に変えるために、ほかの失敗と見分けられなければならないからである。
  メッセージにファイルのパスを含めないのは、503 の応答としてそのまま外に出るため。
  """

  defexception [:model]

  @impl true
  def message(%{model: model}), do: "学習済みモデル #{model} が見つかりません"
end

defmodule GettingStartedMl.Chapter15.Domain do
  @moduledoc """
  第 15 章のドメイン層。予測の入力と、既存のモデルを約束に合わせるアダプター。

  HTTP にも保存の形式にも依存しない。

  モデルそのものには behaviour を作らない。第 8 章の前処理と同じく、モデルは
  「特徴量を受け取って予測を返す関数」でよい。Elixir の無名関数はそのまま値なので、
  包むモジュールを作らずにアダプターが書ける。
  """

  alias GettingStartedMl.{Chapter07, Chapter08}

  @sales_model "cinema"
  @survival_model "survived"
  @survived 1

  @doc "興行収入のモデルの名前。"
  def sales_model, do: @sales_model

  @doc "生存予測のモデルの名前。"
  def survival_model, do: @survival_model

  @doc "モデルの名前を、置き場の順で並べたリスト。ヘルスチェックの並びもこれに従う。"
  def model_names, do: [@sales_model, @survival_model]

  @doc "第 7 章の線形回帰のモデルを、興行収入のモデルの約束（映画 → 数値）に合わせる。"
  def linear_sales_model(model) do
    fn movie ->
      Chapter07.predict_one(model, %{
        SNS1: movie.sns1,
        SNS2: movie.sns2,
        actor: movie.actor,
        original: movie.original
      })
    end
  end

  @doc "第 8 章の学習済みパイプラインを、生存予測のモデルの約束（乗客 → 真偽値）に合わせる。"
  def pipeline_survival_model(pipeline) do
    fn passenger ->
      pipeline
      |> Chapter08.predict(Chapter08.features_table([passenger_row(passenger)]))
      |> hd()
      |> Kernel.==(@survived)
    end
  end

  # 乗客を、第 8 章のパイプラインが読む CSV と同じセルの文字列の行にする。
  # 分からない値は空欄にすると、学習のときに求めた中央値・最頻値で補完される。
  defp passenger_row(passenger) do
    %{
      Pclass: to_string(passenger.pclass),
      Sex: passenger.sex,
      Age: cell(passenger.age),
      SibSp: to_string(passenger.sib_sp),
      Parch: to_string(passenger.parch),
      Fare: to_string(passenger.fare),
      Embarked: cell(passenger.embarked)
    }
  end

  defp cell(nil), do: ""
  defp cell(value), do: to_string(value)
end
